source/common/network/happy_eyeballs_connection_impl.cc

```diff
@@ -0,0 +1,528 @@
+
+  for (auto& connection : connections_) {
+    DUMP_DETAILS(connection);
+  }
+}
+
+ClientConnectionPtr HappyEyeballsConnectionImpl::createNextConnection() {
+  ASSERT(next_address_ < address_list_.size());
+  auto connection = dispatcher_.createClientConnection(
+      address_list_[next_address_++], source_address_,
+      socket_factory_.createTransportSocket(transport_socket_options_), options_);
+  callbacks_wrappers_.push_back(std::make_unique<ConnectionCallbacksWrapper>(*this, *connection));
+  connection->addConnectionCallbacks(*callbacks_wrappers_.back());
+
+  if (per_connection_state_.detect_early_close_when_read_disabled_.has_value()) {
+    connection->detectEarlyCloseWhenReadDisabled(
+        per_connection_state_.detect_early_close_when_read_disabled_.value());
+  }
+  if (per_connection_state_.no_delay_.has_value()) {
+    connection->noDelay(per_connection_state_.no_delay_.value());
+  }
+  if (per_connection_state_.connection_stats_.has_value()) {
+    connection->setConnectionStats(*per_connection_state_.connection_stats_);
+  }
+  if (per_connection_state_.buffer_limits_.has_value()) {
+    connection->setBufferLimits(per_connection_state_.buffer_limits_.value());
+  }
+  if (per_connection_state_.enable_half_close_.has_value()) {
+    connection->enableHalfClose(per_connection_state_.enable_half_close_.value());
+  }
+  if (per_connection_state_.delayed_close_timeout_.has_value()) {
+    connection->setDelayedCloseTimeout(per_connection_state_.delayed_close_timeout_.value());
+  }
+  if (per_connection_state_.start_secure_transport_.has_value()) {
+    ASSERT(per_connection_state_.start_secure_transport_);
+    connection->startSecureTransport();
+  }
+
+  return connection;
+}
+
+void HappyEyeballsConnectionImpl::tryAnotherConnection() {
+  connections_.push_back(createNextConnection());
+  connections_.back()->connect();
+  maybeScheduleNextAttempt();
+}
+
+void HappyEyeballsConnectionImpl::maybeScheduleNextAttempt() {
+  if (next_address_ >= address_list_.size()) {
+    return;
+  }
+  next_attempt_timer_->enableTimer(std::chrono::milliseconds(300));
+}
+
+void HappyEyeballsConnectionImpl::onEvent(ConnectionEvent event,
+                                          ConnectionCallbacksWrapper* wrapper) {
+  wrapper->connection().removeConnectionCallbacks(*wrapper);
+  if (event != ConnectionEvent::Connected) {
+    if (next_address_ < address_list_.size()) {
+      next_attempt_timer_->disableTimer();
+      tryAnotherConnection();
+    }
+    if (connections_.size() > 1) {
+      // Nuke this connection and associated callbacks and let a subsequent attempt proceed.
+      cleanupWrapperAndConnection(wrapper);
+      return;
+    }
+  }
+
+  connect_finished_ = true;
+  next_attempt_timer_->disableTimer();
+
+  // Close and delete up other connections.
+  auto it = connections_.begin();
+  while (it != connections_.end()) {
+    if (it->get() != &(wrapper->connection())) {
+      (*it)->removeConnectionCallbacks(*wrapper);
```

Comment:
can we use ::close() ?

Reply:
close is called on the next line. But close will send events up to the connection callbacks which I think it would be better to avoid?